### ISSUE TITLE
Negative type narrowing fails for protocols with overloaded methods

### ISSUE THREAD

**Issue by BvB93** [python/mypy#12228]
**Bug Report**

Since the merging of https://github.com/python/mypy/pull/9904 it is now possible to use runtime-checkable protocols with overloaded methods. While type narrowing now happens successfully with `isinstance`/`issubclass` checks in the positive case, the same does not hold for the negative case, be it either via `not isinstance`/`not issubclass` or a an `else` statement.

**To Reproduce**
``` python
from typing import overload, Protocol, runtime_checkable
from collections.abc import Sequence

class Foo:
    def func(self, a: int) -> float: ...

class FooOverload:
    @overload
    def func(self, a: int) -> float: ...
    @overload
    def func(self, a: str) -> Sequence[float]: ...

@runtime_checkable
class SupportsFunc(Protocol):
    @overload
    def func(self, a: int) -> float: ...
    @overload
    def func(self, a: str) -> Sequence[float]: ...

foo_union: FooOverload | Foo
```

**Expected Behavior**

``` python

if isinstance(foo_union, SupportsFunc):
    reveal_type(foo_union)  # N: Revealed type is '__main__.FooOverload'
else:
    reveal_type(foo_union)  # N: Revealed type is '__main__.Foo'

```

**Observed Behavior**

``` python

# NOTE: a `not isinstance` check also reveals a `Union`
if isinstance(foo_union, SupportsFunc):
    reveal_type(foo_union)  # N: Revealed type is '__main__.FooOverload'
else:
    reveal_type(foo_union)  # N: Revealed type is 'Union[__main__.FooOverload, __main__.Foo]'

```

**Your Environment**

- Mypy version used: `master` branch (post-https://github.com/python/mypy/commit/b22c4e4a2e9f56480724a45e159552c922af47dc)
- Mypy command-line flags: n.a.
- Mypy configuration options from `mypy.ini` (and other config files): n.a.
- Python version used: 3.10.2
- Operating system and version: Windows 10

**Comment by erictraut** [comment]
This bug was apparently fixed in mypy 0.981. Since then, it produces results consistent with the "expected behavior".

**Comment by hauntsaninja** [comment]
I think fixed by #13303
